### ISSUE TITLE
Inquiry about your CMT paper implementation

### ISSUE THREAD

**Issue by OmarAlsaqa** [yuranusduke/CMT-Convolutional-NN-Meets-ViT#1]
@yuranusduke 
Great work! I have tried my own implementation -[Please check it here](https://github.com/OmarAlsaqa/CMT-Convolutional-Neural-Networks-Meet-Vision-Transformers)- but my model keeps getting NaNs everywhere. Luckily I found yours. I am asking if you have tried to train the model and get results? Just to make sure that the paper has given us enough information to implement it. I looked at the differences between my code and yours and there is not much of a change or that I have made a huge mistake. I am not an expert, I am just a fresh graduate working on improving my skills.

I tried to contact you through e-mail address but gmail doesn't send email to 163.com

**Comment by yuranusduke** [comment]
Hi,
1. I've tried CIFAR10 and CIFAR100 with resized inputs(e.g. 224 x 224) on my single-GPU machine, however, lots of ViTs need large dataset(e.g. ImageNet-21K) to pretrain, I didn't have ideal computational resources, therefore, I trained model with both datasets from scratch. It is obvious that results are not good.(CIFAR10: 77%, CIFAR100: 53%), You may follow the popular training pipeline if you want.
2. I've read your code, I noticed that  in your code, LMHSA first appears NAN, you may debug there, my suggestion is avoiding  
using `permute` and `view` together very often, because sometimes you may get confused. I recommend to use `rearrange` function, which is simple and straightforward.

**Comment by OmarAlsaqa** [comment]
Thank you for your reply.

I will try to use `rearrange` and do more debugging to see what happens but it is nice to know the paper is working.

**Comment by wilile26811249** [comment]
My implement of the CMT, and then training on cifar10, the result is not good (88.16%).  But, I think it need lots of data, because it use the ViT structure.
https://github.com/wilile26811249/CMT_CNN-meet-Vision-Transformer

**Comment by yuranusduke** [comment]
Excellent work! Well, I tried a very small batch size and epochs without advanced augmentation, hope you can work on larger data sets!

**Comment by OmarAlsaqa** [comment]
@wilile26811249 
Great Work, Actually my code is based on your repository but I couldn't get back to it and when I searched about CMT implementations I found @yuranusduke implementation. Thank you both for sharing your implementations.

I do get result using CMT_Ti model but I get NaNs with CMT_XS (I couldn't train using CMT_S  or CMT_B yet due to limitation of computational resources) but it fair to move the issue to your repo :)

@yuranusduke your result is tested on more than a model or just one? Is that one was CMT_Ti?

**Comment by OmarAlsaqa** [comment]
> 
> 
> My implement of the CMT, and then training on cifar10, the result is not good (88.16%). But, I think it need lots of data, because it use the ViT structure.
> https://github.com/wilile26811249/CMT_CNN-meet-Vision-Transformer

I believe 88.16% is a good result as the results given in the paper on CIFAR-10 are fine-tuned with the ImageNet
pretrained checkpoint and trained for 150 epochs.

**Comment by yuranusduke** [comment]
> @wilile26811249
> Great Work, Actually my code is based on your repository but I couldn't get back to it and when I searched about CMT implementations I found @yuranusduke implementation. Thank you both for sharing your implementations.
> 
> I do get result using CMT_Ti model but I get NaNs with CMT_XS (I couldn't train using CMT_S or CMT_B yet due to limitation of computational resources) but it fair to move the issue to your repo :)
> 
> @yuranusduke your result is tested on more than a model or just one? Is that one was CMT_Ti?

I  simply used CMT-Ti model trained with 120 epochs and batch size of 16 on Geforce RTX 2070 for CIFAR10, the accuracy is far from result given in the paper, hope you can work with greater machines to make sure idea indeed works.

**Comment by OmarAlsaqa** [comment]
Currently, The best I can do is to try training on TPUs. If there is any updates I will try to contact you. Thank you for your help. It is fair to close the Issue as I made sure the paper provided the needed information to make it run.
